src/optimizer/OptimizerUtils.cpp

```diff
@@ -32,26 +32,44 @@ Value OptimizerUtils::boundValue(const meta::cpp2::ColumnDef& col,
 Value OptimizerUtils::boundValueWithGT(const meta::cpp2::ColumnDef& col, const Value& v) {
     auto type = SchemaUtil::propTypeToValueType(col.get_type().get_type());
     switch (type) {
-        case Value::Type::INT : {
-            if (v.getInt() == std::numeric_limits<int64_t>::max()) {
-                return v;
+        case Value::Type::INT: {
+            int64_t boundVar;
+            if (v.isFloat()) {
+                // Check if the float value can be cast
+                if (!const_cast<Value&>(v).toInt().second) {
+                    return Value::kNullBadType;
+                }
+                boundVar = const_cast<Value&>(v).toInt().first;
+            } else {
+                boundVar = v.getInt();
+            }
+
+            if (boundVar == std::numeric_limits<int64_t>::max()) {
+                return Value(boundVar);
```

Comment:
I don't think it's a good idea. this leads to the logic more complex in future.
If you want to support value comparison between different type, you’d better cast the data type in validator .